File: twin_econ/product_model.py
from __future__ import annotations

from .params import ScenarioConfig
from .quality_model import quality_score


def module_economics(cfg: ScenarioConfig, modules: int) -> dict[str, float]:
    capped = min(modules, cfg.product.max_modules_per_participant)
    marginal_minutes = capped * cfg.product.module_minutes
    fatigue_contacts = 1 + capped
    turns_total = cfg.cost.avg_scripted_questions + cfg.cost.avg_followups_per_block
    turns_per_minute = turns_total / max(cfg.interview_minutes, 1.0)
    extra_turns_per_module = cfg.product.module_minutes * turns_per_minute
    extra_tokens_out = extra_turns_per_module * cfg.cost.avg_tokens_per_question
    extra_tokens_in = extra_turns_per_module * (cfg.cost.avg_tokens_per_answer + cfg.cost.reflection_update_tokens_per_turn)
    variable_cost_per_module = (
        cfg.product.module_minutes * (cfg.cost.asr_cost_per_minute + cfg.cost.tts_cost_per_minute)
        + (extra_tokens_in / 1000.0) * cfg.cost.price_per_1k_input_tokens
        + (extra_tokens_out / 1000.0) * cfg.cost.price_per_1k_output_tokens
    )
    q_base = quality_score(cfg, cfg.quality_profile, contacts=1)
    q_with_modules = quality_score(
        cfg,
        cfg.quality_profile,
        interview_minutes=cfg.interview_minutes + marginal_minutes,
        contacts=fatigue_contacts,
        domain_complexity_scalar=cfg.product.module_domain_complexity_scalar,
    )
    return {
        "max_modules_per_participant": float(cfg.product.max_modules_per_participant),
        "marginal_minutes_total": marginal_minutes,
        "marginal_cost_per_module": variable_cost_per_module,
        "marginal_quality_gain_per_module": (q_with_modules - q_base) / max(capped, 1),
    }

File: docs-app/src/model/productModel.ts
// Faithful TypeScript port of twin_econ/product_model.py

import type { ScenarioConfig } from './params.ts';
import { qualityScore } from './qualityModel.ts';

export interface ModuleEconomicsResult {
  max_modules_per_participant: number;
  marginal_minutes_total: number;
  marginal_cost_per_module: number;
  marginal_quality_gain_per_module: number;
}

export function moduleEconomics(cfg: ScenarioConfig, modules: number): ModuleEconomicsResult {
  const capped = Math.min(modules, cfg.product.max_modules_per_participant);
  const marginalMinutes = capped * cfg.product.module_minutes;
  const fatigueContacts = 1 + capped;
  const turnsTotal = cfg.cost.avg_scripted_questions + cfg.cost.avg_followups_per_block;
  const turnsPerMinute = turnsTotal / Math.max(cfg.interview_minutes, 1.0);
  const extraTurnsPerModule = cfg.product.module_minutes * turnsPerMinute;
  const extraTokensOut = extraTurnsPerModule * cfg.cost.avg_tokens_per_question;
  const extraTokensIn =
    extraTurnsPerModule * (cfg.cost.avg_tokens_per_answer + cfg.cost.reflection_update_tokens_per_turn);
  const variableCostPerModule =
    cfg.product.module_minutes * (cfg.cost.asr_cost_per_minute + cfg.cost.tts_cost_per_minute) +
    (extraTokensIn / 1000.0) * cfg.cost.price_per_1k_input_tokens +
    (extraTokensOut / 1000.0) * cfg.cost.price_per_1k_output_tokens;

  const qBase = qualityScore(cfg, cfg.quality_profile, undefined, 1);
  const qWithModules = qualityScore(
    cfg,
    cfg.quality_profile,
    cfg.interview_minutes + marginalMinutes,
    fatigueContacts,
    cfg.product.module_domain_complexity_scalar
  );

  return {
    max_modules_per_participant: cfg.product.max_modules_per_participant,
    marginal_minutes_total: marginalMinutes,
    marginal_cost_per_module: variableCostPerModule,
    marginal_quality_gain_per_module: (qWithModules - qBase) / Math.max(capped, 1),
  };
}
